Makefile

```diff
@@ -1,8 +1,12 @@
-SHELL             := /bin/bash
-.SHELLFLAGS       += -e -u -o pipefail
+SHELL       := /bin/bash
+.SHELLFLAGS += -e -u -o pipefail
 
-export IMAGE_NAME := mailserver-testing:ci
-export NAME       ?= $(IMAGE_NAME)
+PARALLEL_JOBS          ?= 2
+export REPOSITORY_ROOT := $(CURDIR)
+export IMAGE_NAME      ?= mailserver-testing:ci
+export NAME            ?= $(IMAGE_NAME)
+
+.PHONY: ALWAYS_RUN
```

Comment:
@casperklein here is what you were looking for :)

Reply:
I am not sure if I am wrong (not an Makefile expert at all😆 ) or you. What is the purpose of ALWAYS_RUN at the moment? 
You cannot run "make ALWAYS_RUN".

For example, I have this:

```
.PHONY: pull start

pull:
        @docker-compose pull

start: pull
		@docker-compose up
```

`pull/start` is phony, because there is no file `pull/start` to build. In the `start` target, `pull` is defined as dependency.

With this I can use `make pull` or `make start` (which also executes `docker-compose pull` before).

That said, I've no clue whats going on with ALWAYS_RUN 😄

Reply:
> What is the purpose of ALWAYS_RUN at the moment? You cannot run "make ALWAYS_RUN".

At the moment, `ALWAYS_RUN` can be provided as a dependency to a target, which will make the target itself behave as if was declared `.PHONY`. `ALWAYS_RUN` itself is not actually a real target and you cannot run it.

> For example, I have this:
> 
> ```
> .PHONY: pull start
> 
> pull:
>         @docker-compose pull
> 
> start: pull
> 		@docker-compose up
> ```
> 
> `pull/start` is phony, because there is no file `pull/start` to build.

A target is not automatically `.PHONY` only because there is not file(s) tracking it, AFAIK. I have a lot of experience with `make`, but I have not read that targets become `.PHONY` implicitly.

> That said, I've no clue whats going on with ALWAYS_RUN

`ALWAYS_RUN` is just a convenience to not declare every target `.PHONY` individually, but have one `.PHONY` "target" (not actually, but `make` does not really care) that you can add as a dependency to make other targets behave as if they were `.PHONY`.

> > This suffices to tell make to always run the target.
> 
> Which target? I am on the go, I'll play/test later a bit.

Every target that has `ALWAYS_RUN` as a dependency. I said _target;; in `make´, this is actually a _recipe_ 😄

Reply:
Here is an example I came up with:

```make
.PHONY: ALWAYS_RUN

main.o: main.c
	@ echo "not phony"
	@ gcc -c $^ -o $@

.PHONY: main1.o
main1.o: main1.c
	@ echo "explicit phony"
	@ gcc -c $^ -o $@

main2.o: main2.c ALWAYS_RUN
	@ echo "implicit phony"
	@ gcc -c main2.c -o $@

clean: ALWAYS_RUN
	@ rm -f main*.o
```

and we have a simple `main.c`, `main1.c` and `main2.c` with `int main() { return 0; }` as content. Now here is me using Make:

```console
georg@swift /tmp  ➜ make
not phony
georg@swift /tmp  ➜ make
make: 'main.o' is up to date.
georg@swift /tmp  ➜ make main1.o
explicit phony
georg@swift /tmp  ➜ make main1.o
explicit phony
georg@swift /tmp  ➜ make main2.o
implicit phony
georg@swift /tmp  ➜ make main2.o
implicit phony
```

and you can see the advantages and disadvantages of `ALWAYS_RUN`. You have it as a dependency, and you can therefore not really use `$^` or `$+` anymore.

Reply:
If you ever want to see or use an over-engineered `Makefile` for compiling C/C++ files with Assembly mixed in:

<details>

```make
SHELL        := /bin/bash -eEu -o pipefail -c

SRCDIR       := src
OUTDIR       := build
DEPDIR       := $(OUTDIR)/dependencies
BINARY       := $(OUTDIR)/binary

CPP          ?= cpp
CC           ?= gcc
CXX          ?= g++
AS           ?= as

SRCFILES     := $(shell find $(SRCDIR) -regextype egrep -regex ".*\.(c|cc|S|asm)")
OBJFILES     := $(addsuffix .o, $(patsubst $(SRCDIR)%, $(OUTDIR)%, $(basename $(SRCFILES))))
DEPFILES     := $(patsubst $(OUTDIR)%.o,$(DEPDIR)%.d,$(OBJFILES))

__CFLAGS     := -Wall -Wextra -pedantic -I $(SRCDIR) -g
CFLAGS       := $(CFLAGS)
CXXFLAGS     := $(CFLAGS)
DEPFLAGS      = -MT $@ -MMD -MP -MF $(DEPDIR)/$*.d

# COMMON TARGETS
# ---------------------------------------------------------

.PHONY: force clean

default: serial_build
all: force
force: clean default

clean:
	@ $(RM) -r $(OUTDIR)

serial_build: $(BINARY)

parallel_build:
	@ $(MAKE) -j $$(nproc) $(BINARY)

$(BINARY): $(OBJFILES)
	@ printf 'LD\t$@\n'
	$(CXX) $(CXXFLAGS) -o $@ $^

# CREATE OBJECT FILES
# ---------------------------------------------------------

# build all .o files from source .c files
$(OUTDIR)/%.o: $(SRCDIR)/%.c | $(OUTDIR) $(DEPDIR)
	@ printf 'CC\t%-40s -> $@\n' '$<'
	@ [[ -d $(@D) ]] || mkdir -p $(@D)
	@ [[ -d $(DEPDIR)/$(*D) ]] || mkdir -p $(DEPDIR)/$(*D)
	@ $(CC) $(DEPFLAGS) $(CFLAGS) -c -o $@ $<

# build all .o files from source .cc files
$(OUTDIR)/%.o: $(SRCDIR)/%.cc | $(OUTDIR) $(DEPDIR)
	@ printf 'CXX\t%-40s -> $@\n' '$<'
	@ [[ -d $(@D) ]] || mkdir -p $(@D)
	@ [[ -d $(DEPDIR)/$(*D) ]] || mkdir -p $(DEPDIR)/$(*D)
	@ $(CXX) $(DEPFLAGS) $(CXXFLAGS) -c -o $@ $<

# build all .o files from source .S files
$(OUTDIR)/%.o: $(SRCDIR)/%.S | $(OUTDIR)
	@ printf 'AS\t%-40s -> $@\n' '$<'
	@ [[ -d $(@D) ]] || mkdir -p $(@D)
	@ $(CPP) -o $(OUTDIR)/$*.clean.S $<
	@ $(AS) -o $@ $(OUTDIR)/$*.clean.S

# build all .o files from source .asm files
$(OUTDIR)/%.o: $(SRCDIR)/%.asm | $(OUTDIR)
	@ printf 'NASM\t%-40s -> $@\n' '$<'
	@ [[ -d $(@D) ]] || mkdir -p $(@D)
	@ nasm -w-zeroing -f elf64 -o $@ $<

# MISCELLANEOUS
# ---------------------------------------------------------

$(DEPDIR) $(OUTDIR) $(QEMU_DIR) $(EFI_BOOT_DIR):
	@- mkdir -p $@

$(DEPFILES):
-include $(wildcard $(DEPFILES))
```
</details>

Reply:
> ALWAYS_RUN is just a convenience to not declare every target .PHONY individually

Ack, now I understand. I always did it as one-liner to achieve the same, e.g.:

`.PHONY: all build generate-accounts backup clean tests lint hadolint shellcheck eclint`